src/sagas/wallet.js

```diff
@@ -0,0 +1,550 @@
+      // the feature flag
+      yield call(featureFlags.ignoreWalletServiceFlag.bind(featureFlags));
+
+      // Restart the whole bundle to make sure we clear all events
+      NativeModules.HTRReloadBundleModule.restart();
+    }
+  }
+
+  walletUtil.storePasswordHash(pin);
+  walletUtil.storeEncryptedWords(words, pin);
+  setKeychainPin(pin);
+
+  yield put(setServerInfo({
+    version: null,
+    network: networkName,
+  }));
+
+  // Wallet might be already ready at this point
+  if (!wallet.isReady()) {
+    const { error } = yield race({
+      success: take(types.WALLET_STATE_READY),
+      error: take(types.WALLET_STATE_ERROR),
+    });
+
+    if (error) {
+      yield put(startWalletFailed());
+      return;
+    }
+  }
+
+  try {
+    yield call(loadTokens);
+  } catch (e) {
+    console.log('Captured error: ', e);
+    yield put(startWalletFailed());
+    return;
+  }
+
+  const featureFlagsThread = yield fork(listenForFeatureFlags, featureFlags);
+
+  yield put(startWalletSuccess());
+
+  // The way the redux-saga fork model works is that if a saga has `forked`
+  // another saga (using the `fork` effect), it will remain active until all
+  // the forks are terminated. You can read more details at
+  // https://redux-saga.js.org/docs/advanced/ForkModel
+  // So, if a new START_WALLET_REQUESTED action is dispatched, we need to cleanup
+  // all attached forks (that will cause the event listeners to be cleaned).
+  while (true) {
+    yield take('START_WALLET_REQUESTED');
+    yield cancel([
+      walletListenerThread,
+      walletReadyThread,
+      featureFlagsThread
+    ]);
+  }
+}
+
+/**
+ * This saga will load both HTR and DEFAULT_TOKEN (if they are different)
+ * and dispatch actions to asynchronously load all registered tokens
+ */
+export function* loadTokens() {
+  const customTokenUid = DEFAULT_TOKEN.uid;
+  const htrUid = hathorLibConstants.HATHOR_TOKEN_CONFIG.uid;
+
+  // Download hathor token balance and history:
+  yield put(tokenFetchBalanceRequested(hathorLibConstants.HATHOR_TOKEN_CONFIG.uid));
+  const { htrBalanceError } = yield race({
+    success: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_SUCCESS, {
+      tokenId: htrUid,
+    })),
+    htrBalanceError: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_FAILED, {
+      tokenId: htrUid,
+    })),
+  });
+  // ...and history:
+  yield put(tokenFetchHistoryRequested(hathorLibConstants.HATHOR_TOKEN_CONFIG.uid));
+  const { htrHistoryError } = yield race({
+    success: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_SUCCESS, {
+      tokenId: htrUid,
+    })),
+    htrHistoryError: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_FAILED, {
+      tokenId: htrUid,
+    })),
+  });
+
+  if (htrBalanceError || htrHistoryError) {
+    throw new Error('Error loading HTR history or balance');
+  }
+
+  if (customTokenUid !== htrUid) {
+    // Download custom token balance:
```

Comment:
The code below here is exactly the same, we could have an aux method for that, what do you think?

Reply:
Yes we should, refactored in 8ffc8fddced0a40aee145d4ab69e68d410b45fe6